Add partial content controller and IO interface

saralein opened the pull request.

saralein committed:
Add partial content controller and IO interface

damonkelley commented on server/src/main/java/com/saralein/server/controller/PartialContentController.java:
```diff
@@ -0,0 +1,134 @@
+    private String formatContentRange(int start, int end, int fileLength) {
+        return String.format("bytes %d-%d/%d", start, end, fileLength);
+    }
+}
```
There are a handful of methods in this class that share a prefix ("parse", "range", "specifies").

saralein replied:
Tried breaking this up into smaller classes based on this.

zdavy commented on server/src/main/java/com/saralein/server/router/Router.java:
```diff
@@ -15,17 +15,20 @@
+    private final Controller partialContentController;
     private final ErrorController errorController;
     private boolean resourceExists = false;
     private boolean resourceIsDirectory = false;
```
So, unrelated to the specific feature, but I was thinking about these resourceExists/resourceIsDirectory properties. They don't seem to be router state, but rather request state. Can we refactor these away from being instance properties and evaluate and use them when they're needed?

saralein replied:
Yeah, I don't think these should be here.  Trying to rework it a bit on [this PR](https://github.com/saralein/java-server/pull/27).

saralein committed:
Break controller into multiple classes

damonkelley commented on server/src/main/java/com/saralein/server/FileHelper.java:
```diff
@@ -38,12 +40,16 @@ public String determineMimeType(String file) {
-    private String removeRootPortionOfPath(Path resource) {
-        return resource.toString().replace(root.toString(), "");
+    public int getFileLength(Path file) throws IOException {
+        return (int) Files.size(file);
```
What does `size` return? Do you need to cast? If it returns a `Integer`, Java should automatically unbox it an cast it to an `int`.

saralein replied:
It returns a long.

damonkelley replied:
🤔 What would happen if the size is larger than `Integer.MAX_VALUE`?

saralein replied:
Yeah, that wouldn't be good.  I'm changing stuff around to long.  Just need to figure out how to use `ByteBuffer.allocation` with longs.  Or whatever I need to do there.

damonkelley commented on server/src/main/java/com/saralein/server/controller/PartialContentController.java:
```diff
@@ -0,0 +1,78 @@
+        } catch (IOException e) {
+            return new Response.Builder()
+                    .status(500)
+                    .build();
```
Do you need to throw a 500, or will the `ErrorController` handle exceptions for you?

saralein replied:
I can use the `ErrorController`.  I wasn't satisfied with it and wanted to rework it, so I think I was avoiding using it.

saralein replied:
`PartialContentController` uses the `ErrorController` now.  But I feel like perhaps this relationship needs to be reworked in a different PR with the `FileController` and `DirectoryController` as well.

damonkelley commented on server/src/main/java/com/saralein/server/partial_content/Range.java:
```diff
@@ -0,0 +1,19 @@
+    public int getEnd() {
+        return end;
+    }
+}
```
👏

damonkelley commented on server/src/main/java/com/saralein/server/router/Router.java:
```diff
@@ -54,6 +57,9 @@ private Controller routeResource(Request request) {
         } else if (isAllowedMethod) {
+            if (requestingPartialContent(request)) {
+                return partialContentController;
+            }
```
I might extract this into a method. As a rule of thumb, I try my hardest to avoid nesting blocks more than one level.

damonkelley commented on server/src/test/java/com/saralein/server/FileHelperTest.java:
```diff
@@ -56,11 +58,14 @@ public void getsMimeTypesOfFiles() {
+    public void getsLengthOfFiles() throws IOException {
+        assertEquals(10, fileHelper.getFileLength(txtPath));
+        assertEquals(15587, fileHelper.getFileLength(jpgPath));
+        assertEquals(2628712, fileHelper.getFileLength(gifPath));
```
Where do these numbers come from?

saralein replied:
Well, the txt file from counting the characters.  But the others are just hanging out in OS info.

damonkelley replied:
Is there a way to get those number programmatically?

damonkelley replied:
As @zacharyjdavy pointed out to me, it is probably not a good idea to get these programmatically. Maybe just using a variable instead of the literal?

damonkelley commented on server/src/main/java/com/saralein/server/controller/PartialContentController.java:
```diff
@@ -0,0 +1,78 @@
+        return validator.isValidRange(range, fileLength)
+                ? getValidRangeResponse(request, file, range, fileLength)
+                : getInvalidRangeResponse(fileLength);
+    }
```
This looks like a pretty complicated method, can we refactor this to make it look a little more clear?

saralein committed:
Rework partial content issues

- Use error controller in partial content controller for 500
- Create method for responses by range validity
- Extract if/else nesting in router to method

saralein committed:
Update file length test

damonkelley commented on server/src/main/java/com/saralein/server/partial_content/RangeParser.java:
```diff
@@ -0,0 +1,37 @@
+public class RangeParser {
+    public Range parse(String range, int fileLength) {
+        String modifiedRange = removeRangeUnit(range);
+        int hyphenIdx = modifiedRange.lastIndexOf("-");
```
What is `Idx`?

saralein replied:
It's index.  Updated the variable name to `hyphenIndex`.

damonkelley commented on server/src/main/java/com/saralein/server/router/Router.java:
```diff
@@ -72,4 +76,15 @@ private Controller routeUri(Request request) {
+            return partialContentController;
+        }
+        return fileController;
+    }
```
This is a little strange to me. Would this still be here after the static middleware PR is completed?

saralein replied:
Modified this so the FileController delegates partial content requests to the PartialContentController.

saralein committed:
Rework file controller to handle partial content vs. router
